backend/src/composite/split.rs

```diff
@@ -43,40 +44,102 @@ pub(crate) fn machine_witness_columns<F: FieldElement>(
     machine_pil: &Analyzed<F>,
     machine_name: &str,
 ) -> Vec<(String, Vec<F>)> {
+    let machine_columns = select_machine_columns(
+        all_witness_columns,
+        machine_pil.committed_polys_in_source_order(),
+    );
+    let size = machine_columns
+        .iter()
+        .map(|(_, column)| column.len())
+        .unique()
+        .exactly_one()
+        .unwrap_or_else(|err| {
+            if err.try_len().unwrap() == 0 {
+                // No witness column, use degree of provided PIL
+                // In practice, we'd at least expect a bus accumulator here, so this should not happen
+                // in any sound setup (after #1498)
+                machine_pil.degree() as usize
+            } else {
+                panic!("Machine {machine_name} has witness columns of different sizes")
+            }
+        });
     let dummy_column_name = format!("{machine_name}.{DUMMY_COLUMN_NAME}");
-    let dummy_column = vec![F::zero(); machine_pil.degree() as usize];
+    let dummy_column = vec![F::zero(); size];
     iter::once((dummy_column_name, dummy_column))
-        .chain(select_machine_columns(
-            all_witness_columns,
-            machine_pil.committed_polys_in_source_order(),
-        ))
+        .chain(machine_columns.into_iter().cloned())
         .collect::<Vec<_>>()
 }
 
 /// Given a set of columns and a PIL describing the machine, returns the fixed column that belong to the machine.
 pub(crate) fn machine_fixed_columns<F: FieldElement>(
-    all_fixed_columns: &[(String, Vec<F>)],
+    all_fixed_columns: &[(String, VariablySizedColumn<F>)],
     machine_pil: &Analyzed<F>,
-) -> Vec<(String, Vec<F>)> {
-    select_machine_columns(
+) -> BTreeMap<usize, Vec<(String, VariablySizedColumn<F>)>> {
+    let machine_columns = select_machine_columns(
         all_fixed_columns,
         machine_pil.constant_polys_in_source_order(),
-    )
+    );
+    let sizes = machine_columns
+        .iter()
+        .map(|(_, column)| column.available_sizes())
+        .collect::<BTreeSet<_>>();
+
+    assert!(
+        sizes.len() <= 1,
+        "All fixed columns of a machine must have the same sizes"
+    );
+
+    sizes
```

Comment:
This is way too nested, isnt there a better way? Flat map or smth?

Reply:
I agree, pretty terrible. Just refactored a bit.